src/analyses/tmpSpecial.ml

```diff
@@ -0,0 +1,151 @@
+(* Analysis that tracks which variables hold the results of calls to math library functions.
+  For each equivalence a set of lvals is tracked, that contains all lvals on which the arguments of the corresponding call depend, so an equivalence can be removed if one of the lvals is written.*)
+
+open GoblintCil
+open Analyses
+
+module Spec =
+struct
+  include Analyses.IdentitySpec
+
+  let name () = "tmpSpecial"
+  module ML = LibraryDesc.MathLifted
+  module LS = SetDomain.ToppedSet(Lval.CilLval) (struct let topname = "All" end)
+  module MlLsProd = Lattice.Prod (ML) (LS)
+  module D = MapDomain.MapBot (Lval.CilLval) (MlLsProd)
+  module C = Lattice.Unit
+
+  let rec resolve (offs : offset) : (CilType.Fieldinfo.t, Basetype.CilExp.t) Lval.offs =
+    match offs with
+    | NoOffset -> `NoOffset
+    | Field (f_info, f_offs) -> `Field (f_info, (resolve f_offs))
+    | Index (i_exp, i_offs) -> `Index (i_exp, (resolve i_offs))
+
+  let ls_of_lv ctx (lval:lval) : LS.t =
+    match lval with
+    | (Var v, offs) -> LS.of_list [(v, resolve offs)]
+    | (Mem e, _) -> (ctx.ask (Queries.MayPointTo e))
+
+  let rec ls_of_exp ctx (exp:exp) : LS.t =
+  match exp with
+  | Const _ -> LS.bot ()
+  | Lval lv -> ls_of_lv ctx lv
+  | SizeOf _ -> LS.bot ()
+  | Real e -> ls_of_exp ctx e
+  | Imag e -> ls_of_exp ctx e
+  | SizeOfE e -> ls_of_exp ctx e
+  | SizeOfStr _ -> LS.empty ()
+  | AlignOf _ -> LS.top () (* TODO: what is this*)
+  | AlignOfE _ -> LS.top () (* TODO: what is this*)
+  | UnOp (_,e,_) -> ls_of_exp ctx e
+  | BinOp (_,e1,e2,_) -> LS.union (ls_of_exp ctx e1) (ls_of_exp ctx e2)
+  | Question (q,e1,e2,_) -> LS.union (ls_of_exp ctx q) (LS.union (ls_of_exp ctx e1) (ls_of_exp ctx e2))
+  | CastE (_,e) -> ls_of_exp ctx e
+  | AddrOf _ -> ctx.ask (Queries.MayPointTo exp)
+  | AddrOfLabel _ -> LS.top () (* TODO: what is this*)
+  | StartOf _ -> LS.top () (* TODO: what is this*)
+
+
+  let context _ _ = ()
+
+  (* transfer functions *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    (* get the set of all lvals written by the assign. Then remove all entrys from the map where the dependencies overlap with the set of written lvals *)
+    let lvalsWritten = ls_of_lv ctx lval in
+    if M.tracing then M.trace "tmpSpecial" "lvalsWritten %a\n" LS.pretty lvalsWritten;
+    D.filter (fun _ (ml, ls) -> LS.is_bot (LS.meet lvalsWritten ls) ) ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    ctx.local
+
+  let enter ctx (lval: lval option) (f:fundec) (args:exp list) : (D.t * D.t) list =
+    [ctx.local, D.bot ()]
+
+  let combine ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (au:D.t) f_ask : D.t =
+    D.bot ()
```

Comment:
Maybe make a comment explaining that we forget all relationships at the call?

Reply:
The analysis can surely be expanded to track relationships interprocedurally. I will look into that (in this PR or in a following one), however, the seemingly most important case, where the result of a function call is put into a tmp variable is covered by the current implementation.

Reply:
Sure, I'm not sure if one even wants to extend it that way, but if we don't we should make this comment here, so people understand it is `D.bot ()` for a good reason!